### PR TITLE
Handle .init_array link_section specially on wasm

### PR DESCRIPTION
**Pull request by ratmice**
Given that wasm-ld now has support for [.init_array](https://github.com/llvm/llvm-project/blob/8f2bd8ae68883592a333f4bdbed9798d66e68630/llvm/lib/MC/WasmObjectWriter.cpp#L1852), it appears we can easily implement that section by falling through to the normal path rather than taking the typical custom_section path for wasm.

The wasm-ld appears to have a bunch of limitations. Only one static with the `link_section` in a crate or else you hit the fatal error in the link above "only one .init_array section fragment supported". They do not get merged.

You can still call multiple constructors by setting it to an array.

```
unsafe extern "C" fn ctor() {
    println!("foo");
}
#[used]
#[link_section = ".init_array"]
static FOO: [unsafe extern "C" fn(); 2] = [ctor, ctor];
```

Another issue appears to be that if crate *A* depends on crate *B*, but *A* doesn't call any symbols from *B* and *B* doesn't `#[export_name = ...]` any symbols, then crate *B*'s constructor will not be called.  The workaround to this is to provide an exported symbol in crate *B*.

### CODE REVIEW

File: compiler/rustc_hir_analysis/src/check/mod.rs [hunk 1]
```diff
@@ -159,21 +159,27 @@ fn maybe_check_static_with_link_section(tcx: TyCtxt<'_>, id: LocalDefId) {
...
     // the consumer's responsibility to ensure all bytes that have been read
     // have defined values.
     if let Ok(alloc) = tcx.eval_static_initializer(id.to_def_id())
         && alloc.inner().provenance().ptrs().len() != 0
     {
-        let msg = "statics with a custom `#[link_section]` must be a \
+        if attrs
+            .link_section
+            .map(|link_section| !link_section.as_str().starts_with(".init_array"))
+            .unwrap_or(true)
```
Comment (ratmice) [inline compiler/rustc_hir_analysis/src/check/mod.rs:176]: I guess this could just be unwrap() because of the early return on line 158-159

**Comment by bjorn3** [comment]
> Another issue appears to be that if crate A depends on crate B, but A doesn't call any symbols from B and B doesn't #[export_name = ...] any symbols, then crate B's constructor will not be called. The workaround to this is to provide an exported symbol in crate B.

Does `#[used]` or `#[used(linker)]` fix that?

**Comment by ratmice** [comment]
> > Another issue appears to be that if crate A depends on crate B, but A doesn't call any symbols from B and B doesn't #[export_name = ...] any symbols, then crate B's constructor will not be called. The workaround to this is to provide an exported symbol in crate B.
> 
> Does `#[used]` or `#[used(linker)]` on the static fix that?

No, neither helps, nor does passing `--no-gc-sections`, I can confirm that the used attributes were present in the `.ll` emitted,
and my guess is that it is a problem in the linker in the generation of the `__wasm_call_ctors` function by the linker.

**Comment by bjorn3** [comment]
And adding `--whole-archive`?

**Comment by ratmice** [comment]
> And adding `--whole-archive`?

Doesn't appear to help either.

I should probably mention if it helps I'm using the wasi-root for the current download version of wasi-sdk.

**Comment by apiraino** [comment]
r? compiler

File: compiler/rustc_hir_analysis/src/check/mod.rs [hunk 2]
```diff
@@ -159,21 +159,27 @@ fn maybe_check_static_with_link_section(tcx: TyCtxt<'_>, id: LocalDefId) {
...
     }
 
-    // For the wasm32 target statics with `#[link_section]` are placed into custom
-    // sections of the final output file, but this isn't link custom sections of
-    // other executable formats. Namely we can only embed a list of bytes,
-    // nothing with provenance (pointers to anything else). If any provenance
-    // show up, reject it here.
+    // For the wasm32 target statics with `#[link_section]` other than `.init_array`
+    // are placed into custom sections of the final output file, but this isn't link
+    // custom sections of other executable formats. Namely we can only embed a list
```
Comment (nnethercote) [inline compiler/rustc_hir_analysis/src/check/mod.rs:164]: Pre-existing, but might as well fix it: "this isn't link custom sections..." is nonsensical phrasing. I'm not sure what the correct words would be.

Comment (bjorn3) [reply compiler/rustc_hir_analysis/src/check/mod.rs:164]: Maybe s/link/like?

**Comment by nnethercote** [comment]
This seems reasonable to me but I don't know much about wasm. @bjorn3, are you willing to review/approve this? Also, is anything from the @bjorn3/@ratmice worth putting into a comment somewhere?

**Comment by ratmice** [comment]
I'm not really sure any of the things discussed are worth putting into a comment, in particular they are limitations of the wasm-ld linker, and could change easily without notice. I had considered putting it into a comment.  But it really feels like something which is just bound to become stale and out of date if at some point in the future if the linker is ever fixed.

It does feel like that stuff could/should be documented somewhere, but perhaps the best place for it to be documented is by linker itself?

**Comment by nnethercote** [comment]
What about a comment like this: "the wasm-ld linker currently has $LIMITATION, so we work around it by doing $X. If this limitation is lifted in the future we can instead do $Y"?  I can imagine somebody in the future finding that comment extremely useful, in a Chesterton's Fence kind of way.

**Comment by ratmice** [comment]
I tried to fix up most of the review comments, and added a new comment like you asked.
I'm still a bit uncertain about really where this comment really belongs, but at least it is *somewhere*, hopefully that is ok.

**Comment by nnethercote** [comment]
The new comment is good, thank you. I will give @bjorn3 a chance to add any extra comments, but if they don't say anything by early next week I think this is ok to merge.

@bors delegate=ratmice

**Comment by ratmice** [comment]
Forgot about this one for a bit, so it has been a bit longer than a week.
Gopefully I understand bors instructions correctly. @bors r=@nnethercote

**Comment by bjorn3** [comment]
This was missing a test. In https://github.com/rust-lang/rust/pull/161862 I've updated the existing `.init_array` test to also work on wasm.

**Pull request merged**